clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -903,8 +903,8 @@ public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
             log.debug("Subscribed to pattern: {}", pattern);
             this.subscriptions.subscribe(pattern, listener);
             this.metadata.needMetadataForAllTopics(true);
-            this.metadata.requestUpdate();
             this.coordinator.updatePatternSubscription(metadata.fetch());
+            this.metadata.requestUpdate();
```

Comment:
@hachikuji Thank you for the reviews. I think there is an issue in the implementation. After the call to `metadata.requestUpdate`, if heartbeat thread does metadata update and clears the update flag, the following line that does `coordinator.updatePatternSubscription(metadata.fetch())` can overwrite the subscription with an older metadata. I think `requestUpdate` should be after `updatePatternSubscription`. This causes an occasional failure in this test, which is fixed with this reordering. Does this make sense?